lib/kino/js.ex

```diff
@@ -0,0 +1,362 @@
+  JavaScript side only receives the initial data and there is no
+  further interaction with the Elixir side. To introduce such
+  interaction, see `Kino.JS.Live` as a next step in our discussion.
+  '''
+
+  defstruct [:module, :data]
+
+  @type t :: %__MODULE__{module: module(), data: term()}
+
+  defmacro __using__(opts) do
+    quote location: :keep, bind_quoted: [opts: opts] do
+      import Kino.JS, only: [asset: 2]
+
+      @before_compile Kino.JS
+
+      Module.register_attribute(__MODULE__, :inline_assets, accumulate: true)
+
+      assets_path = opts[:assets_path]
+      asset_paths = Kino.JS.__paths__(assets_path)
+
+      loaded_assets =
+        for path <- asset_paths do
+          abs_path = Path.join(assets_path, path)
+          @external_resource Path.relative_to_cwd(abs_path)
+          content = File.read!(abs_path)
+          {path, content}
+        end
+
+      @loaded_assets loaded_assets
+      @assets_path assets_path
+
+      # Force recompilation if new assets are added
+      def __mix_recompile__? do
+        current_paths = Kino.JS.__paths__(unquote(assets_path))
+        :erlang.md5(current_paths) != unquote(:erlang.md5(asset_paths))
+      end
+    end
+  end
+
+  @doc ~S'''
+  Defines an asset file.
+
+  This serves as a convenience when prototyping or building
+  simple widgets, otherwise you most likely want to put assets
+  in separate files. See the [Assets](#module-assets) for more details.
+
+  ## Examples
+
+      asset "main.js" do
+        """
+        export function init(ctx, data) {
+          ...
+        }
+        """
+      end
+
+      asset "main.css" do
+        """
+        .box {
+          ...
+        }
+        """
+      end
+  '''
+  defmacro asset(name, do: block) do
+    quote bind_quoted: [name: name, content: block] do
+      Module.put_attribute(__MODULE__, :inline_assets, {name, content})
+    end
+  end
+
+  defmacro __before_compile__(env) do
+    inline_assets = Module.get_attribute(env.module, :inline_assets)
+    loaded_assets = Module.get_attribute(env.module, :loaded_assets)
+    assets_path = Module.get_attribute(env.module, :assets_path)
+
+    any_inline_assets? = inline_assets != []
+    assets_path_defined? = assets_path != nil
+
+    assets =
+      case {any_inline_assets?, assets_path_defined?} do
+        {true, false} ->
+          inline_assets
+
+        {false, true} ->
+          if loaded_assets == [] do
+            IO.warn(
+              "assets directory specified for #{inspect(env.module)}, but no files" <>
+                " found in #{inspect(assets_path)}",
+              Macro.Env.stacktrace(env)
+            )
+          end
+
+          loaded_assets
+
+        {true, true} ->
+          IO.warn(
+            "ignoring files in #{inspect(assets_path)} because #{inspect(env.module)}" <>
+              " already defines inline assets with the assets/2 macro",
+            Macro.Env.stacktrace(env)
+          )
```

Comment:
Is there a reason for not supporting both given how it is implemented?

Reply:
Nothing in particular, it's mostly for consistency, as there's no reason to use both?

Reply:
Your call!